update

Denisska1231 opened the pull request.

Denisska1231 committed:
update

Denisska1231 committed:
commit

Quercussi reviewed (changes requested):
Please fix these errors.
1. Assigning `nextpc` to `pc + 1` should be executed before calling the decoder, as the simulator will run address 0 twice.
2. Apology for confusion, but please replace `setPc()` with `setNextpc()` and amend its instructions as named. Remember, the type instruction doesn't always call `getPc()`; there are certain instructions that Stage has to execute directly, such as assigning `nextpc = pc + 1`.  This method exists in the first place to let BEQ and JALR instructions to set the next execution address.
3. It needn't check `!isHalt` in the `iterator()`, as it is already checked in `simlutate()`.
4. Add `getInstruction()` method for Decoder please.
5. Don't forget to call the decoder.
6. Also Don't forget to remove main method.

Quercussi commented:
I forgot to mention this earlier: please also call `printState()` method as required in the instructor's project description.

Denisska1231 committed:
add getInstruction() and fix code

Quercussi commented on src/Stage.java:
```diff
@@ -6,18 +6,20 @@ public class Stage {
         nextPc = 0;
         isHalt = false;
+        decoder = Decoder.getInstance();
     }
```
Please add another field `stepCount` that dictates how many steps have been iterated. `stepCount` should be incremented at the end of `iterate()` method.

Quercussi commented on src/Stage.java:
```diff
@@ -62,26 +62,14 @@ public int getInstructionCount() {
         }
         return instructionCount;
```
1. Please replace `int getInstructionCount()` method with `void setInstructionCount()` instead.
2. Create a new field `instructionCount` for `printState()` to use.

Quercussi commented on src/Stage.java:
```diff
@@ -26,13 +28,11 @@ public void iterate() {
             pc = nextPc;
-            nextPc = pc + 1;
-        }
+            decoder.execute(this);
```
The instruction must be executed before setting `pc` to `nextPc`.

Quercussi reviewed (changes requested):
Please fix these

Denisska1231 committed:
- add stepCount
- replace int getInstructionCount() method with void setInstructionCount()
- Create a new field instructionCount
- fix instruction calling

Quercussi commented on src/Stage.java:
```diff
@@ -25,10 +27,11 @@ public void iterate() {
             // Increment the program counter to the next instruction
             pc = nextPc;
-            decoder.execute(this);
+            instructionCount++;
```
As intended, instruction Count will not be iterated, but I'll merge your pull-request for now.

Quercussi commented:
Wow! You actually forgot to call `printState()`.
